### PR TITLE
Workflow versioning

### PR DESCRIPTION
**Pull request by marten**
Adds a new API call which will store the current workflow attributes as a discrete WorkflowVersion object, and mark that as the currently published version. This includes the associated workflow contents as well.

When calling the read actions for workflows, if given `published=true` as a parameter, Panoptes will serialize the published `WorkflowVersion` instead of the `Workflow` itself. This means that a project owner can publish a workflow, and then continue editing. Those new edits will not be seen by volunteers until the project owner publishes again.

I'm thinking about the getting Panoptes to always return the latest Workflow instead of the published WorkflowVersion if a project is set to "development" instead of "live".

I want to make sure that it won't be easy for project owners to forget to Publish. Any thoughts on this?

Follow-up tasks for PFE:

* Send `published=true` when loading workflows for the classify interface.
* But don't send that if a project owner came there by clicking `Test workflow`, since they'll want to test out their changes.
* Add a "Publish changes" button in the project builder to call Panoptes' new API.
* Allow owners to edit workflows even while a project is live.

WorkflowVersions are standard AR models, and distinct from Papertrail's versioning system. The main reason for this is that Papertrail makes it hard to retrieve a specific version, and slow to reify it. Papertrail will likely be removed long-term, but since this involves figuring out how to handle classifications referring to a specific version of a workflow, I decided to do that in a later PR. The workflow_version column in classifications still uses the Papertrail versions. I'm aware that this will cause some confusion, and any thoughts on how to alleviate would be much appreciated.

# TODO

- [x] Make versions in a before_save rather than on publish
- [ ] Remove publish action, instead rely on links?
- [ ] Link classifications against workflow version -> this is actually done by PFE, not doing this for now

# Review checklist

- [ ] First, the most important one: is this PR small enough that you can actually review it? Feel free to just reject a branch if the changes are hard to review due to the length of the diff.
- [ ] If there are any migrations, will they the previous version of the app work correctly after they've been run (e.g. the don't remove columns still known about by ActiveRecord).
- [ ] If anything changed with regards to the public API, are those changes also documented in the `apiary.apib` file?
- [ ] Are all the changes covered by tests? Think about any possible edge cases that might be left untested.

### CODE REVIEW

**Comment by marten** [comment]
@camallen no need to review the code yet, but could you have a glance at the PR description and see if there's anything in there you think might not be a good idea? Some of it is still todo in this PR.

**Comment by camallen** [comment]
Translations currently copy the strings from the `*_content` models (if they exist) or just the resource attributes. Those originating strings can come from anywhere but will involve some code changes to the translation strings extractor https://github.com/zooniverse/Panoptes/blob/db02a12f073348abac5dfb76485c676babec160c/app/models/concerns/translation_strings.rb#L30

I have an old PR to move the `has_many` *_content models to a `has_one` that should help set the stage for a way to move the primary language strings from the contents back to the originating resource and we can delete the linked content models - https://github.com/zooniverse/Panoptes/pull/2754

>  The workflow_version column in classifications still uses the Papertrail versions. I'm aware that this will cause some confusion, and any thoughts on how to alleviate would be much appreciated.

I'm not 100% how to deal with version numbering of the old data and it was a concern in the above PR as well. Maybe migrating the paper trail data out to your new data model and having an seperate column with those version numbers to allow the old data to be found. Though i think that model schema doesn't capture the same data. Our main consideration is for the classification data exports and the fidelity of that data. Also I think we can drop the version routes for these resources as they aren't used and are buggy.

> Send published=true when loading workflows for the classify interface

I think of the published state as locking the representation and i'd default to serializing this data for the workflow resource. It could be controlled via a query param to modify that behaviour.

>I'm thinking about the getting Panoptes to always return the latest Workflow instead of the published WorkflowVersion if a project is set to "development" instead of "live".

This may be confusing and complicates state management. Would this be better controlled via a query parameter on the client?

> I want to make sure that it won't be easy for project owners to forget to Publish. Any thoughts on this?

Not sure - probably UI signalling that there is a discrepancy and a warning / check when setting the live state.

**Comment by srallen** [comment]
> This may be confusing and complicates state management. Would this be better controlled via a query parameter on the client?

Query parameter is a good idea and could be added to the URL from the 'test this workflow' button in the lab. We could add text there to make it explicit that this is how you can test the latest in development workflow.

**Comment by marten** [comment]
>>Send published=true when loading workflows for the classify interface
>
>I think of the published state as locking the representation and i'd default to serializing this data for the workflow resource. It could be controlled via a query param to modify that behaviour.

I considered that, but I think the downsides are pretty big. It's very surprising to an API consumer when they post changes and never see them if they query the resource normally.

File: app/serializers/workflow_serializer.rb [hunk 1]
```diff
@@ -35,11 +35,19 @@ def content_version
...
+    requested_version.tasks
+  end
+
+  def first_task
+    requested_version.first_task
+  end
+
+  def requested_version
+    if @context[:published]
+      @model.published_version || @model
```
[the remark this answers is not in the document]
Comment (marten) [reply app/serializers/workflow_serializer.rb]: I could point it at `@model.published_version || WorkflowVersion.build_from(@model)` to make both the same type?

**Comment by marten** [comment]
> Nice work - combined with the current lab auto save this will litter a lot of workflow versions. What's the plan to backfill the new models from say paper trail versions so we can still use the dumps?

Soon. I've seen an occasional inconsistency between the old ModelVersion and the new columns on `workflow`. Once that's fixed we can do away with the Papertrail ones and convert everything over.

> WorkflowVersions.(major|minor)_version seems like redundancy in the naming, perhaps major|minor_number 2/10 on this.

Nice one, it was nagging me a little as well and I like that much better.

> I assume that the published_version_id column will end up as non-nullable in a later PR?

Can't really do that I think, since you get a circular dependency. Then you need a workflow_id for the WorkflowVersion, need a published_version_id for the Workflow.

> The docker-compose changes could come in an isolated PR and maybe replace cellect with designator?

Sorry, I should've dropped those from here, esp redis shouldn't be committed.

File: docker-compose.yml [hunk 2]
```diff
@@ -3,30 +3,10 @@ postgres:
   environment:
     - "POSTGRES_USER=panoptes"
     - "POSTGRES_PASSWORD=panoptes"
-  ports:
```
[the remark this answers is not in the document]
Comment (marten) [reply docker-compose.yml]: Will remove from this PR, but could we move it to a different port? With it as-is, if someone happens to have a local PG, they can't boot the docker version.

File: config/routes.rb [hunk 3]
```diff
@@ -118,13 +118,16 @@
...
-      json_api_resources :workflows, links: [:subject_sets, :retired_subjects, :tutorials], versioned: true do
+      json_api_resources :workflows, links: [:subject_sets, :retired_subjects, :tutorials, :workflow_versions, :published_version], versioned: true do
         media_resources :attached_images, classifications_export: { except: [:create] }
 
         post "/retired_subjects", to: "workflows#retire_subjects"
         post "/classifications_export", to: "workflows#create_classifications_export", format: false
+        post '/publish', to: 'workflows#publish'
       end
 
+      json_api_resources :workflow_versions, links: [:workflow]
```
[the remark this answers is not in the document]
Comment (marten) [reply config/routes.rb]: No, in my mind WorkflowVersions aren't mutable in any way.

File: app/serializers/workflow_serializer.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (camallen) [reply app/serializers/workflow_serializer.rb]: Up to you. i'd be happy have some comment to say why these two models work the same via duck typing for future selves. If you think it's cheap enough to build the version model then it seems more consistent.

File: docker-compose.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (camallen) [reply docker-compose.yml]: different port works for me, i assume we don't specify the port numbers in the getting started docs.

File: app/models/workflow.rb [hunk 4]
```diff
@@ -59,6 +69,8 @@ def self.translatable_attributes
   end
 
   before_save :update_version
+  after_create :save_version
```
[the remark this answers is not in the document]
Comment (marten) [reply app/models/workflow.rb]: I considered that but figured I'd stick with doing it the same way as Papertrail is doing for now.

**Pull request merged**